File: desktop/qbot/memory_retrieval.py
from __future__ import annotations

import json
import re
import unicodedata
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Mapping, Sequence

from sqlalchemy import func, select, text
from sqlalchemy.exc import SQLAlchemyError

from qbot.persistence.database import Database
from qbot.persistence.memory import MEMORY_SCOPES, MemoryRecord
from qbot.persistence.tables import memories


class MemoryRetrievalError(ValueError):
    pass


@dataclass(frozen=True, slots=True)
class MemoryQuery:
    scopes: tuple[str, ...]
    text: str = ""
    owner_id: str | None = None
    conversation_id: str | None = None
    task_id: str | None = None
    entities: tuple[str, ...] = ()
    min_trust: float = 0.0
    limit: int = 10


@dataclass(frozen=True, slots=True)
class MemoryScore:
    total_points: int
    keyword_milli: int
    entity_milli: int
    recency_milli: int
    importance_milli: int
    trust_milli: int

    @property
    def total(self) -> float:
        return self.total_points / 100_000


@dataclass(frozen=True, slots=True)
class MemoryHit:
    record: MemoryRecord
    score: MemoryScore


@dataclass(frozen=True, slots=True)
class MemoryScoringWeights:
    keyword: int = 40
    entity: int = 20
    recency: int = 15
    importance: int = 10
    trust: int = 15

    def validate(self) -> None:
        values = (
            self.keyword,
            self.entity,
            self.recency,
            self.importance,
            self.trust,
        )
        if any(value < 0 for value in values):
            raise MemoryRetrievalError("memory scoring weights must be non-negative")
        if sum(values) != 100:
            raise MemoryRetrievalError("memory scoring weights must sum to 100")


_WORD_RE = re.compile(r"[0-9a-z]+|[\u3400-\u4dbf\u4e00-\u9fff]+")
_OWNER_SCOPES = {"USER_PERSONA", "CONTACT_PROFILE"}
_FTS_SIGNAL_RE = re.compile(r"^[0-9a-z]{3,}$")
_FTS_TABLE = "memory_search_fts"


class MemoryRetriever:
    """Deterministic read-only retrieval over authoritative promoted memory.

    This module freezes filtering and scoring semantics before any FTS or
    embedding-backed candidate accelerator is introduced. Future indexes may
    reduce the candidate set, but must not change eligibility or ranking.
    """

    def __init__(
        self,
        database: Database,
        *,
        weights: MemoryScoringWeights | None = None,
        enable_fts: bool = True,
    ) -> None:
        self.database = database
        self.weights = weights or MemoryScoringWeights()
        self.weights.validate()
        self.enable_fts = bool(enable_fts)
        self._fts_disabled = False
        self.last_candidate_backend = "scan"

    def retrieve(
        self,
        query: MemoryQuery,
        *,
        now: datetime | None = None,
    ) -> list[MemoryHit]:
        scopes = self._validate_query(query)
        instant = self._normalize_now(now or datetime.now(UTC))
        terms = _terms(query.text)
        query_entities = _normalized_unique(query.entities)
        requires_relevance = bool(terms or query_entities)

        candidate_ids = self._fts_candidate_ids(
            terms=terms,
            query_entities=query_entities,
        )
        statement = (
            select(memories)
            .where(
                memories.c.state == "PROMOTED",
                memories.c.scope.in_(scopes),
                memories.c.trust >= query.min_trust,
            )
            .order_by(memories.c.memory_id.asc())
        )
        if candidate_ids is not None:
            self.last_candidate_backend = "fts5"
            if not candidate_ids:
                return []
            statement = statement.where(
                memories.c.memory_id.in_(candidate_ids)
            )
        else:
            self.last_candidate_backend = "scan"

        with self.database.engine.connect() as conn:
            rows = conn.execute(statement).mappings().all()

        hits: list[MemoryHit] = []
        for row in rows:
            if not self._domain_matches(row, query):
                continue
            if not self._is_temporally_valid(row, instant):
                continue

            record = _record(row)
            keyword_milli = self._keyword_score(record, terms)
            entity_milli = self._entity_score(record, query_entities)
            if requires_relevance and keyword_milli == 0 and entity_milli == 0:
                continue

            recency_milli = self._recency_score(record, instant)
            importance_milli = _unit_to_milli(
                record.importance if record.importance is not None else 0.5
            )
            trust_milli = _unit_to_milli(record.trust)
            total = (
                keyword_milli * self.weights.keyword
                + entity_milli * self.weights.entity
                + recency_milli * self.weights.recency
                + importance_milli * self.weights.importance
                + trust_milli * self.weights.trust
            )
            hits.append(
                MemoryHit(
                    record=record,
                    score=MemoryScore(
                        total_points=total,
                        keyword_milli=keyword_milli,
                        entity_milli=entity_milli,
                        recency_milli=recency_milli,
                        importance_milli=importance_milli,
                        trust_milli=trust_milli,
                    ),
                )
            )

        hits.sort(key=self._sort_key)
        return hits[: query.limit]

    def _fts_candidate_ids(
        self,
        *,
        terms: tuple[str, ...],
        query_entities: tuple[str, ...],
    ) -> set[str] | None:
        if not self.enable_fts or self._fts_disabled or self.database.safe_mode:
            return None

        signals = self._fts_signals(terms, query_entities)
        if signals is None:
            return None

        try:
            with self.database.engine.begin() as conn:
                conn.execute(
                    text(
                        "CREATE VIRTUAL TABLE IF NOT EXISTS "
                        f"{_FTS_TABLE} USING fts5("
                        "memory_id UNINDEXED, searchable, tokenize='trigram'"
                        ")"
                    )
                )

                memory_count = conn.execute(
                    select(func.count()).select_from(memories)
                ).scalar_one()
                index_count = conn.execute(
                    text(f"SELECT COUNT(*) FROM {_FTS_TABLE}")
                ).scalar_one()

                if int(memory_count) != int(index_count):
                    self._rebuild_fts(conn)

                expression = " OR ".join(
                    f'"{signal}"' for signal in signals
                )
                rows = conn.execute(
                    text(
                        f"SELECT memory_id FROM {_FTS_TABLE} "
                        "WHERE searchable MATCH :expression"
                    ),
                    {"expression": expression},
                ).scalars().all()
            return {str(memory_id) for memory_id in rows}
        except SQLAlchemyError:
            self._fts_disabled = True
            return None

    @staticmethod
    def _fts_signals(
        terms: tuple[str, ...],
        query_entities: tuple[str, ...],
    ) -> tuple[str, ...] | None:
        signals = tuple(dict.fromkeys((*terms, *query_entities)))
        if not signals:
            return None
        if any(_FTS_SIGNAL_RE.fullmatch(signal) is None for signal in signals):
            return None
        return signals

    @staticmethod
    def _rebuild_fts(conn) -> None:
        conn.execute(text(f"DELETE FROM {_FTS_TABLE}"))
        rows = conn.execute(
            select(
                memories.c.memory_id,
                memories.c.content,
                memories.c.entities_json,
            ).order_by(memories.c.memory_id.asc())
        ).mappings().all()
        for row in rows:
            raw_entities = json.loads(str(row["entities_json"] or "[]"))
            if not isinstance(raw_entities, list):
                raw_entities = []
            searchable = _normalize_text(
                " ".join(
                    (
                        str(row["content"]),
                        *(str(value) for value in raw_entities),
                    )
                )
            )
            conn.execute(
                text(
                    f"INSERT INTO {_FTS_TABLE}(memory_id, searchable) "
                    "VALUES (:memory_id, :searchable)"
                ),
                {
                    "memory_id": str(row["memory_id"]),
                    "searchable": searchable,
                },
            )

    @staticmethod
    def _validate_query(query: MemoryQuery) -> tuple[str, ...]:
        if not query.scopes:
            raise MemoryRetrievalError("at least one memory scope is required")
        scopes = tuple(dict.fromkeys(scope.strip().upper() for scope in query.scopes))
        unknown = sorted(set(scopes) - MEMORY_SCOPES)
        if unknown:
            raise MemoryRetrievalError(
                "unsupported memory scopes: " + ", ".join(unknown)
            )
        if any(scope in _OWNER_SCOPES for scope in scopes) and not query.owner_id:
            raise MemoryRetrievalError(
                "owner_id is required for USER_PERSONA/CONTACT_PROFILE retrieval"
            )
        if "CONVERSATION_MEMORY" in scopes and not query.conversation_id:
            raise MemoryRetrievalError(
                "conversation_id is required for CONVERSATION_MEMORY retrieval"
            )
        if "TASK_MEMORY" in scopes and not query.task_id:
            raise MemoryRetrievalError(
                "task_id is required for TASK_MEMORY retrieval"
            )
        if not 0.0 <= float(query.min_trust) <= 1.0:
            raise MemoryRetrievalError("min_trust must be between 0 and 1")
        if query.limit <= 0 or query.limit > 100:
            raise MemoryRetrievalError("limit must be between 1 and 100")
        return scopes

    @staticmethod
    def _domain_matches(row: Mapping[str, object], query: MemoryQuery) -> bool:
        scope = str(row["scope"])
        if scope in _OWNER_SCOPES:
            return row["owner_id"] == query.owner_id
        if scope == "CONVERSATION_MEMORY":
            return row["conversation_id"] == query.conversation_id
        if scope == "TASK_MEMORY":
            return row["task_id"] == query.task_id
        return scope == "SYSTEM_POLICY"

    @staticmethod
    def _is_temporally_valid(
        row: Mapping[str, object],
        now: datetime,
    ) -> bool:
        valid_from_raw = row["valid_from"]
        valid_to_raw = row["valid_to"]
        if valid_from_raw is not None:
            valid_from = _parse_timestamp(str(valid_from_raw))
            if valid_from is None or now < valid_from:
                return False
        if valid_to_raw is not None:
            valid_to = _parse_timestamp(str(valid_to_raw))
            if valid_to is None or now >= valid_to:
                return False
        return True

    @staticmethod
    def _keyword_score(record: MemoryRecord, terms: tuple[str, ...]) -> int:
        if not terms:
            return 0
        searchable = _normalize_text(
            " ".join((record.content, *record.entities))
        )
        matched = sum(1 for term in terms if term in searchable)
        return matched * 1000 // len(terms)

    @staticmethod
    def _entity_score(
        record: MemoryRecord,
        query_entities: tuple[str, ...],
    ) -> int:
        if not query_entities:
            return 0
        record_entities = set(_normalized_unique(record.entities))
        matched = sum(1 for entity in query_entities if entity in record_entities)
        return matched * 1000 // len(query_entities)

    @staticmethod
    def _recency_score(record: MemoryRecord, now: datetime) -> int:
        anchor = _parse_timestamp(record.valid_from) if record.valid_from else None
        if anchor is None:
            anchor = _parse_timestamp(record.created_at)
        if anchor is None:
            return 0

        age_seconds = max(0.0, (now - anchor).total_seconds())
        age_days = age_seconds / 86_400
        if age_days <= 1:
            return 1000
        if age_days <= 7:
            return 850
        if age_days <= 30:
            return 700
        if age_days <= 180:
            return 450
        if age_days <= 365:
            return 250
        return 100

    @staticmethod
    def _sort_key(hit: MemoryHit) -> tuple[object, ...]:
        created = _parse_timestamp(hit.record.created_at)
        created_stamp = created.timestamp() if created is not None else float("-inf")
        return (
            -hit.score.total_points,
            -hit.score.keyword_milli,
            -hit.score.entity_milli,
            -hit.score.recency_milli,
            -hit.score.importance_milli,
            -hit.score.trust_milli,
            -created_stamp,
            hit.record.memory_id,
        )

    @staticmethod
    def _normalize_now(value: datetime) -> datetime:
        if value.tzinfo is None or value.utcoffset() is None:
            raise MemoryRetrievalError("now must be timezone-aware")
        return value.astimezone(UTC)


def _normalize_text(value: str) -> str:
    normalized = unicodedata.normalize("NFKC", str(value)).casefold()
    return " ".join(normalized.split())


def _terms(value: str) -> tuple[str, ...]:
    normalized = _normalize_text(value)
    return tuple(dict.fromkeys(_WORD_RE.findall(normalized)))


def _normalized_unique(values: Sequence[str]) -> tuple[str, ...]:
    result: list[str] = []
    seen: set[str] = set()
    for raw in values:
        value = _normalize_text(raw)
        if not value or value in seen:
            continue
        seen.add(value)
        result.append(value)
    return tuple(result)


def _unit_to_milli(value: float) -> int:
    number = min(1.0, max(0.0, float(value)))
    return int(number * 1000 + 0.5)


def _parse_timestamp(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None
    if parsed.tzinfo is None or parsed.utcoffset() is None:
        return None
    return parsed.astimezone(UTC)


def _record(row: Mapping[str, object]) -> MemoryRecord:
    raw_entities = json.loads(str(row["entities_json"] or "[]"))
    if not isinstance(raw_entities, list):
        raw_entities = []
    return MemoryRecord(
        memory_id=str(row["memory_id"]),
        schema_version=str(row["schema_version"]),
        state=str(row["state"]),
        scope=str(row["scope"]),
        owner_id=_optional_string(row["owner_id"]),
        conversation_id=_optional_string(row["conversation_id"]),
        task_id=_optional_string(row["task_id"]),
        content=str(row["content"]),
        entities=tuple(str(value) for value in raw_entities),
        importance=_optional_float(row["importance"]),
        trust=float(row["trust"]),
        confidence=_optional_float(row["confidence"]),
        source_type=str(row["source_type"]),
        source_message_id=_optional_string(row["source_message_id"]),
        source_event_id=_optional_string(row["source_event_id"]),
        valid_from=_optional_string(row["valid_from"]),
        valid_to=_optional_string(row["valid_to"]),
        supersedes=_optional_string(row["supersedes"]),
        created_at=str(row["created_at"]),
    )


def _optional_string(value: object) -> str | None:
    return None if value is None else str(value)


def _optional_float(value: object) -> float | None:
    return None if value is None else float(value)

File: android/app/src/main/java/dev/qbot/android/data/MemoryRetrieval.kt
package dev.qbot.android.data

import android.icu.lang.UCharacter
import androidx.sqlite.db.SupportSQLiteDatabase
import dev.qbot.android.data.db.MemoryEntity
import dev.qbot.android.data.db.QbotDatabase
import java.text.Normalizer
import java.time.Duration
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.Locale
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONArray

private val RETRIEVAL_MEMORY_SCOPES = setOf(
    "SYSTEM_POLICY",
    "USER_PERSONA",
    "CONTACT_PROFILE",
    "CONVERSATION_MEMORY",
    "TASK_MEMORY",
)

private val OWNER_SCOPES = setOf(
    "USER_PERSONA",
    "CONTACT_PROFILE",
)

private val TERM_PATTERN = Regex(
    "[0-9a-z]+|[\\u3400-\\u4dbf\\u4e00-\\u9fff]+",
)

private val FTS_SIGNAL_PATTERN = Regex("^[0-9a-z]{3,}$")
private val ASCII_RUN_PATTERN = Regex("[0-9a-z]+")
private const val MEMORY_FTS_TABLE = "memory_search_fts"

class MemoryRetrievalError(message: String) : IllegalArgumentException(message)

data class MemoryQuery(
    val scopes: List<String>,
    val text: String = "",
    val ownerId: String? = null,
    val conversationId: String? = null,
    val taskId: String? = null,
    val entities: List<String> = emptyList(),
    val minTrust: Double = 0.0,
    val limit: Int = 10,
)

data class MemoryScore(
    val totalPoints: Int,
    val keywordMilli: Int,
    val entityMilli: Int,
    val recencyMilli: Int,
    val importanceMilli: Int,
    val trustMilli: Int,
) {
    val total: Double
        get() = totalPoints / 100_000.0
}

data class MemoryHit(
    val record: MemoryRecord,
    val score: MemoryScore,
)

data class MemoryScoringWeights(
    val keyword: Int = 40,
    val entity: Int = 20,
    val recency: Int = 15,
    val importance: Int = 10,
    val trust: Int = 15,
) {
    fun validate() {
        val values = listOf(
            keyword,
            entity,
            recency,
            importance,
            trust,
        )
        if (values.any { it < 0 }) {
            throw MemoryRetrievalError(
                "memory scoring weights must be non-negative",
            )
        }
        if (values.sum() != 100) {
            throw MemoryRetrievalError(
                "memory scoring weights must sum to 100",
            )
        }
    }
}

class MemoryRetriever(
    private val database: QbotDatabase,
    private val weights: MemoryScoringWeights = MemoryScoringWeights(),
    private val enableFts: Boolean = true,
) {
    private val dao = database.qbotDao()
    private val ftsMutex = Mutex()
    private var ftsDisabled = false

    var lastCandidateBackend: String = "scan"
        private set

    init {
        weights.validate()
    }

    suspend fun retrieve(
        query: MemoryQuery,
        now: Instant = Instant.now(),
    ): List<MemoryHit> {
        val scopes = validateQuery(query)
        val terms = terms(query.text)
        val queryEntities = normalizedUnique(query.entities)
        val requiresRelevance = terms.isNotEmpty() || queryEntities.isNotEmpty()

        val candidateIds = ftsCandidateIds(
            terms = terms,
            queryEntities = queryEntities,
        )
        val candidates =
            if (candidateIds == null) {
                lastCandidateBackend = "scan"
                dao.memoriesByState("PROMOTED")
            } else {
                lastCandidateBackend = "fts4"
                if (candidateIds.isEmpty()) {
                    emptyList()
                } else {
                    dao.memoriesByStateAndIds(
                        state = "PROMOTED",
                        memoryIds = candidateIds.toList().sorted(),
                    )
                }
            }

        val hits = mutableListOf<MemoryHit>()
        for (entity in candidates) {
            if (entity.scope !in scopes) continue
            if (entity.trust < query.minTrust) continue
            if (!domainMatches(entity, query)) continue
            if (!isTemporallyValid(entity, now)) continue

            val record = record(entity)
            val keywordMilli = keywordScore(record, terms)
            val entityMilli = entityScore(record, queryEntities)
            if (
                requiresRelevance &&
                keywordMilli == 0 &&
                entityMilli == 0
            ) {
                continue
            }

            val recencyMilli = recencyScore(record, now)
            val importanceMilli = unitToMilli(record.importance ?: 0.5)
            val trustMilli = unitToMilli(record.trust)
            val totalPoints =
                keywordMilli * weights.keyword +
                    entityMilli * weights.entity +
                    recencyMilli * weights.recency +
                    importanceMilli * weights.importance +
                    trustMilli * weights.trust

            hits += MemoryHit(
                record = record,
                score = MemoryScore(
                    totalPoints = totalPoints,
                    keywordMilli = keywordMilli,
                    entityMilli = entityMilli,
                    recencyMilli = recencyMilli,
                    importanceMilli = importanceMilli,
                    trustMilli = trustMilli,
                ),
            )
        }

        return hits
            .sortedWith(hitComparator)
            .take(query.limit)
    }

    private suspend fun ftsCandidateIds(
        terms: List<String>,
        queryEntities: List<String>,
    ): Set<String>? {
        if (!enableFts || ftsDisabled) return null

        val signals = ftsSignals(terms, queryEntities) ?: return null
        return ftsMutex.withLock {
            if (ftsDisabled) return@withLock null
            try {
                val db = database.openHelper.writableDatabase
                ensureFtsTable(db)

                val memoryCount = scalarCount(
                    db,
                    "SELECT COUNT(*) FROM memories",
                )
                val indexCount = scalarCount(
                    db,
                    "SELECT COUNT(*) FROM $MEMORY_FTS_TABLE",
                )
                if (memoryCount != indexCount) {
                    rebuildFts(db)
                }

                val expression = signals.joinToString(" OR ") { signal ->
                    val grams = trigrams(signal)
                    grams.joinToString(
                        separator = " AND ",
                        prefix = "(",
                        postfix = ")",
                    ) { gram -> "\"$gram\"" }
                }

                val result = linkedSetOf<String>()
                db.query(
                    "SELECT memory_id FROM $MEMORY_FTS_TABLE " +
                        "WHERE grams MATCH ?",
                    arrayOf(expression),
                ).use { cursor ->
                    val idIndex = cursor.getColumnIndexOrThrow("memory_id")
                    while (cursor.moveToNext()) {
                        result += cursor.getString(idIndex)
                    }
                }
                result
            } catch (_: Exception) {
                ftsDisabled = true
                null
            }
        }
    }

    private fun ftsSignals(
        terms: List<String>,
        queryEntities: List<String>,
    ): List<String>? {
        val signals = linkedSetOf<String>()
        signals += terms
        signals += queryEntities
        if (signals.isEmpty()) return null
        if (signals.any { FTS_SIGNAL_PATTERN.matches(it).not() }) {
            return null
        }
        return signals.toList()
    }

    private fun ensureFtsTable(db: SupportSQLiteDatabase) {
        db.execSQL(
            "CREATE VIRTUAL TABLE IF NOT EXISTS $MEMORY_FTS_TABLE " +
                "USING fts4(memory_id, grams, tokenize=simple)",
        )
    }

    private fun rebuildFts(db: SupportSQLiteDatabase) {
        db.beginTransaction()
        try {
            db.execSQL("DELETE FROM $MEMORY_FTS_TABLE")
            db.query(
                "SELECT memory_id, content, entities_json " +
                    "FROM memories ORDER BY memory_id",
            ).use { cursor ->
                val idIndex = cursor.getColumnIndexOrThrow("memory_id")
                val contentIndex = cursor.getColumnIndexOrThrow("content")
                val entitiesIndex =
                    cursor.getColumnIndexOrThrow("entities_json")
                while (cursor.moveToNext()) {
                    val memoryId = cursor.getString(idIndex)
                    val content = cursor.getString(contentIndex)
                    val entities = decodeEntities(
                        cursor.getString(entitiesIndex),
                    )
                    val searchable = normalizeText(
                        buildList {
                            add(content)
                            addAll(entities)
                        }.joinToString(" "),
                    )
                    val grams = gramDocument(searchable)
                    db.execSQL(
                        "INSERT INTO $MEMORY_FTS_TABLE" +
                            "(memory_id, grams) VALUES (?, ?)",
                        arrayOf(memoryId, grams),
                    )
                }
            }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    private fun gramDocument(value: String): String =
        ASCII_RUN_PATTERN
            .findAll(value)
            .flatMap { match ->
                trigrams(match.value).asSequence()
            }
            .distinct()
            .joinToString(" ")

    private fun trigrams(value: String): List<String> {
        if (value.length < 3) return emptyList()
        return buildList {
            for (index in 0..value.length - 3) {
                add(value.substring(index, index + 3))
            }
        }.distinct()
    }

    private fun scalarCount(
        db: SupportSQLiteDatabase,
        sql: String,
    ): Int =
        db.query(sql).use { cursor ->
            if (!cursor.moveToFirst()) 0 else cursor.getInt(0)
        }

    private fun validateQuery(query: MemoryQuery): Set<String> {
        if (query.scopes.isEmpty()) {
            throw MemoryRetrievalError(
                "at least one memory scope is required",
            )
        }
        val scopes = linkedSetOf<String>()
        query.scopes.forEach { raw ->
            scopes += raw.trim().uppercase(Locale.ROOT)
        }

        val unknown = scopes - RETRIEVAL_MEMORY_SCOPES
        if (unknown.isNotEmpty()) {
            throw MemoryRetrievalError(
                "unsupported memory scopes: " +
                    unknown.sorted().joinToString(", "),
            )
        }
        if (
            scopes.any { it in OWNER_SCOPES } &&
            query.ownerId.isNullOrBlank()
        ) {
            throw MemoryRetrievalError(
                "owner_id is required for " +
                    "USER_PERSONA/CONTACT_PROFILE retrieval",
            )
        }
        if (
            "CONVERSATION_MEMORY" in scopes &&
            query.conversationId.isNullOrBlank()
        ) {
            throw MemoryRetrievalError(
                "conversation_id is required for " +
                    "CONVERSATION_MEMORY retrieval",
            )
        }
        if (
            "TASK_MEMORY" in scopes &&
            query.taskId.isNullOrBlank()
        ) {
            throw MemoryRetrievalError(
                "task_id is required for TASK_MEMORY retrieval",
            )
        }
        if (query.minTrust !in 0.0..1.0) {
            throw MemoryRetrievalError(
                "min_trust must be between 0 and 1",
            )
        }
        if (query.limit !in 1..100) {
            throw MemoryRetrievalError(
                "limit must be between 1 and 100",
            )
        }
        return scopes
    }

    private fun domainMatches(
        entity: MemoryEntity,
        query: MemoryQuery,
    ): Boolean =
        when (entity.scope) {
            "USER_PERSONA",
            "CONTACT_PROFILE",
            -> entity.ownerId == query.ownerId

            "CONVERSATION_MEMORY" ->
                entity.conversationId == query.conversationId

            "TASK_MEMORY" ->
                entity.taskId == query.taskId

            else -> entity.scope == "SYSTEM_POLICY"
        }

    private fun isTemporallyValid(
        entity: MemoryEntity,
        now: Instant,
    ): Boolean {
        entity.validFrom?.let { raw ->
            val validFrom = parseTimestamp(raw) ?: return false
            if (now < validFrom) return false
        }
        entity.validTo?.let { raw ->
            val validTo = parseTimestamp(raw) ?: return false
            if (now >= validTo) return false
        }
        return true
    }

    private fun keywordScore(
        record: MemoryRecord,
        terms: List<String>,
    ): Int {
        if (terms.isEmpty()) return 0
        val searchable = normalizeText(
            buildList {
                add(record.content)
                addAll(record.entities)
            }.joinToString(" "),
        )
        val matched = terms.count { it in searchable }
        return matched * 1000 / terms.size
    }

    private fun entityScore(
        record: MemoryRecord,
        queryEntities: List<String>,
    ): Int {
        if (queryEntities.isEmpty()) return 0
        val recordEntities = normalizedUnique(record.entities).toSet()
        val matched = queryEntities.count { it in recordEntities }
        return matched * 1000 / queryEntities.size
    }

    private fun recencyScore(
        record: MemoryRecord,
        now: Instant,
    ): Int {
        val anchor = record.validFrom?.let(::parseTimestamp)
            ?: parseTimestamp(record.createdAt)
            ?: return 0

        val duration = Duration.between(anchor, now)
        val ageSeconds = maxOf(
            0.0,
            duration.seconds + duration.nano / 1_000_000_000.0,
        )
        val ageDays = ageSeconds / 86_400.0
        return when {
            ageDays <= 1.0 -> 1000
            ageDays <= 7.0 -> 850
            ageDays <= 30.0 -> 700
            ageDays <= 180.0 -> 450
            ageDays <= 365.0 -> 250
            else -> 100
        }
    }

    private val hitComparator =
        compareByDescending<MemoryHit> { it.score.totalPoints }
            .thenByDescending { it.score.keywordMilli }
            .thenByDescending { it.score.entityMilli }
            .thenByDescending { it.score.recencyMilli }
            .thenByDescending { it.score.importanceMilli }
            .thenByDescending { it.score.trustMilli }
            .thenByDescending {
                parseTimestamp(it.record.createdAt) ?: Instant.MIN
            }
            .thenBy { it.record.memoryId }

    private fun record(entity: MemoryEntity): MemoryRecord =
        MemoryRecord(
            memoryId = entity.memoryId,
            schemaVersion = entity.schemaVersion,
            state = entity.state,
            scope = entity.scope,
            ownerId = entity.ownerId,
            conversationId = entity.conversationId,
            taskId = entity.taskId,
            content = entity.content,
            entities = decodeEntities(entity.entitiesJson),
            importance = entity.importance,
            trust = entity.trust,
            confidence = entity.confidence,
            sourceType = entity.sourceType,
            sourceMessageId = entity.sourceMessageId,
            sourceEventId = entity.sourceEventId,
            validFrom = entity.validFrom,
            validTo = entity.validTo,
            supersedes = entity.supersedes,
            createdAt = entity.createdAt,
        )

    private fun decodeEntities(value: String): List<String> {
        val array = JSONArray(value)
        return buildList {
            for (index in 0 until array.length()) {
                add(array.getString(index))
            }
        }
    }

    private fun terms(value: String): List<String> {
        val normalized = normalizeText(value)
        return TERM_PATTERN
            .findAll(normalized)
            .map { it.value }
            .distinct()
            .toList()
    }

    private fun normalizedUnique(values: List<String>): List<String> {
        val seen = linkedSetOf<String>()
        values.forEach { raw ->
            val value = normalizeText(raw)
            if (value.isNotEmpty()) seen += value
        }
        return seen.toList()
    }

    private fun normalizeText(value: String): String {
        val normalized = Normalizer.normalize(
            value,
            Normalizer.Form.NFKC,
        )
        val folded = UCharacter.foldCase(normalized, true)
        val result = StringBuilder()
        var pendingSpace = false
        for (char in folded) {
            val isWhitespace =
                Character.isWhitespace(char) ||
                    Character.isSpaceChar(char) ||
                    char == '\u0085'
            if (isWhitespace) {
                if (result.isNotEmpty()) pendingSpace = true
                continue
            }
            if (pendingSpace) {
                result.append(' ')
                pendingSpace = false
            }
            result.append(char)
        }
        return result.toString()
    }

    private fun unitToMilli(value: Double): Int {
        val number = value.coerceIn(0.0, 1.0)
        return (number * 1000 + 0.5).toInt()
    }

    private fun parseTimestamp(value: String): Instant? =
        runCatching {
            OffsetDateTime
                .parse(value)
                .withOffsetSameInstant(ZoneOffset.UTC)
                .toInstant()
        }.getOrNull()
}
